hudi-common/src/main/java/org/apache/hudi/common/util/ClusteringUtils.java

```diff
@@ -0,0 +1,183 @@
+      }
+      HoodieRequestedReplaceMetadata requestedReplaceMetadata = TimelineMetadataUtils.deserializeRequestedReplaceMetadta(content.get());
+      if (WriteOperationType.CLUSTER.name().equals(requestedReplaceMetadata.getOperationType())) {
+        return Option.of(Pair.of(requestedReplaceInstant, requestedReplaceMetadata.getClusteringPlan()));
+      }
+      return Option.empty();
+    } catch (IOException e) {
+      throw new HoodieIOException("Error reading clustering plan " + requestedReplaceInstant.getTimestamp(), e);
+    }
+  }
+
+  /**
+   * Get filegroups to pending clustering instant mapping for all pending clustering plans.
+   * This includes all clustering operattions in 'requested' and 'inflight' states.
+   */
+  public static Map<HoodieFileGroupId, HoodieInstant> getAllFileGroupsInPendingClusteringPlans(
+      HoodieTableMetaClient metaClient) {
+    Stream<Pair<HoodieInstant, HoodieClusteringPlan>> pendingClusteringPlans = getAllPendingClusteringPlans(metaClient);
+    Stream<Map.Entry<HoodieFileGroupId, HoodieInstant>> resultStream = pendingClusteringPlans.flatMap(clusteringPlan ->
+        // get all filegroups in the plan
+        getFileGroupEntriesInClusteringPlan(clusteringPlan.getLeft(), clusteringPlan.getRight()));
+
+    Map<HoodieFileGroupId, HoodieInstant> resultMap = resultStream.collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    LOG.info("Found " + resultMap.size() + " files in pending clustering operations");
+    return resultMap;
+  }
+
+  public static Stream<Pair<HoodieFileGroupId, HoodieInstant>> getFileGroupsInPendingClusteringInstant(
+      HoodieInstant instant, HoodieClusteringPlan clusteringPlan) {
+    Stream<HoodieFileGroupId> partitionToFileIdLists = clusteringPlan.getInputGroups().stream().flatMap(ClusteringUtils::getFileGroupsFromClusteringGroup);
+    return partitionToFileIdLists.map(e -> Pair.of(e, instant));
+  }
+
+  private static Stream<Map.Entry<HoodieFileGroupId, HoodieInstant>> getFileGroupEntriesInClusteringPlan(
+      HoodieInstant instant, HoodieClusteringPlan clusteringPlan) {
+    return getFileGroupsInPendingClusteringInstant(instant, clusteringPlan).map(entry ->
+        new AbstractMap.SimpleEntry<>(entry.getLeft(), entry.getRight()));
+  }
+
+  private static Stream<HoodieFileGroupId> getFileGroupsFromClusteringGroup(HoodieClusteringGroup group) {
+    return group.getSlices().stream().map(slice -> new HoodieFileGroupId(slice.getPartitionPath(), slice.getFileId()));
+  }
+
+  /**
+   * Create clustering plan from input fileSliceGroups.
+   */
+  public static HoodieClusteringPlan createClusteringPlan(String strategyClassName,
+                                                          Map<String, String> strategyParams,
+                                                          List<FileSlice>[] fileSliceGroups) {
+    List<HoodieClusteringGroup> clusteringGroups = new ArrayList<>();
+    for (int i = 0; i < fileSliceGroups.length; i++) {
```

Comment:
Can we change this loop to java 8 stream style, fileSliceGroups.stream().foreach()..

Reply:
Sure. will fix it.